From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Wed, 26 Mar 2025 01:46:49 +0300
Subject: [PATCH] Leaf: Improve BlockEntity ticking isRemoved check


diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 7eb6da13dc47eaeac6e70d4e2935c1cc022c6400..d2eed7a0cf0c2c9dbcfb272cf89194f11d37151c 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -992,13 +992,26 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
 
     public static class RebindableTickingBlockEntityWrapper implements TickingBlockEntity {
         public TickingBlockEntity ticker;
+        private @Nullable BlockEntity blockEntityReference = null; // DivineMC - Improve BlockEntity ticking isRemoved check
 
         RebindableTickingBlockEntityWrapper(TickingBlockEntity ticker) {
             this.ticker = ticker;
+            // DivineMC start - Improve BlockEntity ticking isRemoved check
+            if (ticker instanceof BoundTickingBlockEntity<?> boundTicker) {
+                blockEntityReference = boundTicker.blockEntity;
+            }
+            // DivineMC end - Improve BlockEntity ticking isRemoved check
         }
 
         public void rebind(TickingBlockEntity ticker) {
             this.ticker = ticker;
+            // DivineMC start - Improve BlockEntity ticking isRemoved check
+            if (ticker instanceof BoundTickingBlockEntity<?> boundTicker) {
+                blockEntityReference = boundTicker.blockEntity;
+            } else {
+                blockEntityReference = null;
+            }
+            // DivineMC end - Improve BlockEntity ticking isRemoved check
         }
 
         @Override
@@ -1008,6 +1021,12 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
 
         @Override
         public boolean isRemoved() {
+            // DivineMC start - Improve BlockEntity ticking isRemoved check
+            if (blockEntityReference != null) {
+                return blockEntityReference.isRemoved();
+            }
+            // DivineMC end - Improve BlockEntity ticking isRemoved check
+
             return this.ticker.isRemoved();
         }
 
