From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Mon, 3 Mar 2025 01:16:29 +0300
Subject: [PATCH] Extend Sound API


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index d309e28d7e6756afed82390fd2a99a6131e29140..468e2fa5e059e19c14d095c840723f10a6aa9287 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -817,4 +817,29 @@ public interface Block extends Metadatable, Translatable, net.kyori.adventure.tr
         return this.getBlockData().getDestroySpeed(itemStack, considerEnchants);
     }
     // Paper end - destroy speed API
+
+    // DivineMC start - Extend Sound API
+    /**
+     * Plays a sound at the location of the block
+     *
+     * @param sound sound to play
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        emitSound(sound, org.bukkit.SoundCategory.BLOCKS, volume, pitch);
+    }
+
+    /**
+     * Plays a sound at the location of the block
+     *
+     * @param sound sound to play
+     * @param category category of the sound
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(getLocation().toCenterLocation(), sound, category, volume, pitch);
+    }
+    // DivineMC end - Extend Sound API
 }
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 97ba2c698da6820eae7d254008579bb64052f474..5a8635890b83b5eab24a3c155b98b50d07e9d7e4 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1269,4 +1269,35 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     void setImmuneToFire(@Nullable Boolean fireImmune);
     // Purpur end - Fire Immunity API
+
+    // DivineMC start - Extend Sound API
+    /**
+     * Plays a sound at the location of the entity
+     *
+     * @param sound sound to play
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, float volume, float pitch) {
+        org.bukkit.SoundCategory soundGroup = switch (this) {
+            case HumanEntity humanEntity -> org.bukkit.SoundCategory.PLAYERS;
+            case Ambient ambient -> org.bukkit.SoundCategory.AMBIENT;
+            case Monster monster -> org.bukkit.SoundCategory.HOSTILE;
+            default -> org.bukkit.SoundCategory.NEUTRAL;
+        };
+        emitSound(sound, soundGroup, volume, pitch);
+    }
+
+    /**
+     * Plays a sound at the location of the block
+     *
+     * @param sound sound to play
+     * @param category category of the sound
+     * @param volume volume of the sound
+     * @param pitch pitch of the sound
+     */
+    default void emitSound(@NotNull org.bukkit.Sound sound, @NotNull org.bukkit.SoundCategory category, float volume, float pitch) {
+        getWorld().playSound(this, sound, category, volume, pitch);
+    }
+    // DivineMC end - Extend Sound API
 }
