From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sat, 11 Jan 2025 23:22:12 +0300
Subject: [PATCH] Optimize default values for configs


diff --git a/src/main/java/org/purpurmc/purpur/PurpurConfig.java b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
index 3dc90c1b194e3c313a80d6ed741ba98d238dc318..d23062496466f46a01c4ef297a740febb134c090 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
@@ -240,7 +240,7 @@ public class PurpurConfig {
         laggingThreshold = getDouble("settings.lagging-threshold", laggingThreshold);
     }
 
-    public static boolean useAlternateKeepAlive = false;
+    public static boolean useAlternateKeepAlive = true; // DivineMC - Optimize default values for configs
     private static void useAlternateKeepAlive() {
         useAlternateKeepAlive = getBoolean("settings.use-alternate-keepalive", useAlternateKeepAlive);
     }
diff --git a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
index d4a6cf2fec1125d853c5d7181cbec861834b3b55..9ef7fce5b8352ef6dd21601fa600eaa5f50295a7 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
@@ -3055,8 +3055,8 @@ public class PurpurWorldConfig {
     public boolean villagerDisplayTradeItem = true;
     public int villagerSpawnIronGolemRadius = 0;
     public int villagerSpawnIronGolemLimit = 0;
-    public int villagerAcquirePoiSearchRadius = 48;
-    public int villagerNearestBedSensorSearchRadius = 48;
+    public int villagerAcquirePoiSearchRadius = 16; // DivineMC - Optimize default values for configs
+    public int villagerNearestBedSensorSearchRadius = 16; // DivineMC - Optimize default values for configs
     private void villagerSettings() {
         villagerRidable = getBoolean("mobs.villager.ridable", villagerRidable);
         villagerRidableInWater = getBoolean("mobs.villager.ridable-in-water", villagerRidableInWater);
